### ISSUE TITLE
Define correct attribute for Content Warnings

### ISSUE THREAD

**Issue by ThisIsMissEm** [swicg/activitypub-trust-and-safety#1]
Taken originally from: https://github.com/w3c/activitystreams/issues/583#

Currently mastodon (and inspired/compatible fediverse software) use the summary and summaryMap properties for content warnings, which isn't ideal since these are also used for actual summaries of Articles, Documents and other object types.

We need to define a new property or set of properties specifically for the purposes of doing content warnings and content labeling.

Content Warnings can largely be just a duplication of what we have for summary/summaryMap (i.e., contentWarning/contentWarningMap), as it's just a formalisation of an existing practice.

**Comment by snarfed** [comment]
Starting with a bang, real practical utility here. I like it!

**Comment by bumblefudge** [comment]
Here's the only chatter I could find about formalizing `as:sensitive`:

https://github.com/w3c/activitystreams/issues/571#issuecomment-1870888419

https://socialhub.activitypub.rocks/t/property-under-tag/3488/19

where else should I be looking?

**Comment by bumblefudge** [comment]
I am exactly the opposite of an ontology queen, but my intuition is that sensitive should be a property that nests multiple possible subproperties, since few communities would have a category of the sensitive fully congruous/coextensive with any other's, and would probably want to focus filtering efforts on only some sensitives and not others. whatever shape the ontological thinking goes, i'm hugely supportive of migrating from `summary` to any form of `sensitive` that can be spec'd out and making `summary` summary again 💪

**Comment by ThisIsMissEm** [comment]
@bumblefudge https://swicg.github.io/miscellany/#sensitive

**Comment by AnSavvides** [comment]
> I am exactly the opposite of an ontology queen, but my intuition is that sensitive should be a property that nests multiple possible subproperties

To @bumblefudge's point, it'd be great for us to discuss / explore if we could go a step further beyond mirroring `summary`/`summaryMap`, and instead introduce a canonical set of sub-properties, along with the ability to add custom ones for more flexibility. Maybe something along these lines?

```
contentWarning: { violence: boolean, nudity: boolean, ..., other: string }
```

That would mean that a piece of content could have multiple warnings, and a `custom` property for anything that doesn't fit whatever list of subproperties this is seeded with (which is probably a whole separate exercise in and of itself).

Though to @evanp's point in https://github.com/w3c/activitystreams/issues/583#issuecomment-1969476912:

> One complication is how we shift from using summary for content warnings to using contentWarning or whatever property you think up is deprecating the use of summary for CWs. The problem is that if clients don't use it, there's some potential for bad experience for the user -- seeing something they don't want to see. We'll need to design that transition carefully; I don't have a clear picture of it right now.

I expect mirroring `summary`/`summaryMap` would mean an easier transition from one to the other, so probably a path to adoption and effort we'd require folks to support this should probably be a consideration.

**Comment by ThisIsMissEm** [comment]
@AnSavvides I think there's two or three things here: 
- content labelling (i.e., violence, nudity, gore, politics, etc)
- content summary (user wants to collapse their post behind some text)
- content warning (a textual description that accompanies content labelling explaining why)

Looks like I didn't create the issue to track content labelling yet

**Comment by AnSavvides** [comment]
> The immediate issue with a canonical list is that there is huge cultural variance between what is considered sensitive or not.
> It's IMHO best to allow for distinct and/or overlapping vocabularies in order to avoid being prescriptive.

Great points, and very much see what you are saying around cultural variance and not being prescriptive. I'll have a read through #15 and follow up there!

**Comment by tesaguri** [comment]
As for the problem of <q>canonical</q> labels, I think labels can be `@id`s rather than a fixed set of strings, so that spec-defined labels, community-provided shared labels and even ad-hoc content warnings can coexist, like so:

```json
{
  "type": "Note",
  "ex:moderationLabel": [
    "ex:violentContent",
    "https://example.com/label/42",
    {
      "type": "Note",
      "name": "Spoiler of the latest episode"
    }
  ]
}
```

An unobvious point of this approach is that, while multiple "labels" for a single post make sense, you might not usually want multiple "content warnings", so maybe content warnings should be separated from labels to make it easy to restrict their cardinality. But there may be cases where you want an author-provided warning and a moderator-provided warning at the same time, so… I'm not sure if it's a good idea to restrict cardinality of content warnings either.

**Comment by jfinkhaeuser** [comment]
@tesaguri That's the essence of #15, though that issue distinguishes between defining and referencing a vocabulary. You could of course always include individual terms immediately as you describe. Maybe that's a good additional thought for that other issue, too.

**Comment by ThisIsMissEm** [comment]
This is a rather interesting article on Content Warnings and labeling of content: https://shelraphen.com/on-content-warnings/

**Comment by evanp** [comment]
`as:sensitive` is defined in the Miscellany report:

https://swicg.github.io/miscellany/

**Comment by ThisIsMissEm** [comment]
@evanp thanks for the update, have added that in the issue description; that report was still being written when i wrote this issue, hence saying the work was in progress.

**Comment by ThisIsMissEm** [comment]
After discussion with a new participant about the T&S taskforce, I shared a few directions that a FEP for this could go in, where at its most basic, it's just copying `summary`/`summaryMap` (the first two examples below), or we could go for something that allows more flexibility, like allowing a different actor to apply the content warning (e.g., the moderation team applying the content warning — intent here isn't for arbitrarily adding annotations to content, that'd be #4 )

I like the idea of defining a `ContentWarning` type and `contentWarning` property that contains the warning, beyond just a simple string. Both paths allow for localisation of the content warning through a `*Map` property, so `contentWarningMap` or just `contentMap` in the case of a `ContentWarning` object type. However, using a fuller object allows for things that we can't currently do within the constrains of the `summary` property.

The FEP should include the different use cases for summary vs content warning vs content labeling (specifically calling out that the FEP is **not** addressing labeling or annotations).

The most basic form of a content warning property would be a direct copy of `summary` / `summaryMap`, would look like the following:
```json
{
  "@context": "https://www.w3.org/ns/activitystreams",
  "attributedTo": "https://social.example/jane",
  "content": "Something adult oriented, which I'm obviously omitting here",
  "type": "Note",
  "contentWarning": "Contains adult content, mentions of sex"
}
```

Or with localisation:
```json
{
  "@context": "https://www.w3.org/ns/activitystreams",
  "attributedTo": "https://social.example/jane",
  "content": "Something adult oriented, which I'm obviously omitting here",
  "type": "Note",
  "contentWarningMap": {
    "en": "Contains adult content, mentions of sex",
    "de": "Enthält Inhalte für Erwachsene, Erwähnungen von Sex"
  }
}
```

### Using an Object to wrap the `contentWarning`

Self-authored content warning, but using an object like a Note within the Note to contain additional metadata about the content warning, the `ContentWarning` type would just extend `Note`, so `Note` could definitely just be reused here:
```json
{
  "@context": "https://www.w3.org/ns/activitystreams",
  "attributedTo": "https://social.example/jane",
  "content": "Something adult oriented, which I'm obviously omitting here",
  "type": "Note",
  "contentWarning": {
    "type": "ContentWarning",
    "attributedTo": "https://social.example/jane",
    "content": "Contains adult content, mentions of sex"
  }
}
```

Changing attribution of the content warning, to allow moderators to apply content warnings to posts not by them, but by people for whom they are responsible for the moderation of their content:
```json
{
  "@context": "https://www.w3.org/ns/activitystreams",
  "attributedTo": "https://social.example/jane",
  "content": "Something adult oriented, which I'm obviously omitting here",
  "type": "Note",
  "contentWarning": {
    "type": "ContentWarning",
    "attributedTo": "https://social.example/moderators",
    "content": "Contains adult content, mentions of sex"
  }
}
```

**Comment by ChaosExAnima** [comment]
Hello, aforementioned new participant here! One question I have is how we'd want to handle multiple editors of CWs? Such as a user putting one on and then moderators adding to it.

Would that be making `attributedTo` an array, just showing the last editor, or is this something that would be more up to implementers?

**Comment by ThisIsMissEm** [comment]
> One question I have is how we'd want to handle multiple editors of CWs? Such as a user putting one on and then moderators adding to it.

I'm not sure that's something this should handle? That sounds more like a Moderator annotating a post (e.g., to correct misinformation) which isn't the intended use of this

**Comment by ClearlyClaire** [comment]
I am not entirely convinced by a `ContentWarning` object, especially if it extends `Note` (this far exceeds our current usage, and you could have Content Warnings with Content Warnings…?).

`contentWarning` and `contentWarningMap` totally make sense to me.

I would just suggest those to be defined as text and *not* HTML.

**Comment by renchap** [comment]
Something that needs to be in the FEP: how to support the transition period where some software will support reading the CWs from the new attributes, but others will still consider `summary` as a CW.

I am not sure what the behaviour here should be but we need to think about:
- authoring software that still uses `summary`
- consuming software that does not know about the new CW attribute. I suspect this should be up to the authoring software to choose for how long they want to still send the CW in the `summary` field, and decide when "enough of the ecosystem has support for it" so they stop doing so
- authoring software supporting the new attribute, and explicitly wants to use the `summary` as not-a-CW. This could be done by having `contentWarning: {}` or similar in the `Note`, which would mean "I support the `contentWarning` attribute, and I chose to add a `summary`, please do not treat it as a CW`", so that software supporting this can behave correctly.

I suspect that the transition period will last a few years, and as CWs are related to T&S we need to be careful about respecting this transition period so all users have time to support it and wont face unexpected un-CWed posts.

**Comment by ShadowJonathan** [comment]
@renchap one more scenario; a software uses `contentWarning`, but also uses `summary` to provide fallback behaviour to a large part of the fediverse at that moment

It's unclear how a software should interpret `summary` then, would there need to be an extra key that says "no, summary is used differently", or how could a defacto transition period be handled like this?

**Comment by ThisIsMissEm** [comment]
I think the general order of things is that `contentWarning` is more significant than `summary`, so if both are present, then show as a content warning, don't show summary. Arguably you could design a UI that includes both `contentWarning` and `summary`, but folks would need to know to be careful not to put anything triggering in the `summary`.

**Comment by evanp** [comment]
> Here's the only chatter I could find about formalizing `as:sensitive`:

https://swicg.github.io/miscellany/#sensitive

**Comment by ThisIsMissEm** [comment]
So there had actually been a progress an decision here, which is that the content warning text should actually be in the `summary`, and the fact it's a content warning is indicated via the `sensitive` attribute, and we were going to draft a FEP in accordance with that.

The rationale was that in long form there'd probably be no useful difference between a `contentWarning` property and `summary`, and that using content labelling using shared vocabularies is likely more meaningful for filtering instead of matching contents of free text fields.

**Comment by ThisIsMissEm** [comment]
@evanp I don't see how `dcterm:subject` would apply or give anything meaningful here, unless you were to tie everything back to wikidata or something? (RIP Freebase)

We'd probably recommend use `tags` with elements of type Label (#84)

**Comment by evanp** [comment]
`dc:subject` is the subject of the content.

http://purl.org/dc/terms/subject

It can be an URI or a literal value. Using literal values lets the author define subjects like "nudity", "US politics", and/or "personal finance". This is usually what people are looking for in a content warning.

A similar property is `schema:about`.

https://schema.org/about

It's more directly for describing another object.

I'll take a look at `label` also. A literal for describing the content seems like a very generic term, so I think we should try to use an existing vocabulary before we define a new one.

**Comment by evanp** [comment]
I opened #87 for this topic. `summary` is simply not available for formalisation as a content warning property.

**Comment by ThisIsMissEm** [comment]
> `dc:subject` is the subject of the content.
> 
> http://purl.org/dc/terms/subject
> 
> It can be an URI or a literal value. Using literal values lets the author define subjects like "nudity", "US politics", and/or "personal finance". This is usually what people are looking for in a content warning.
> 
> A similar property is `schema:about`.
> 
> https://schema.org/about
> 
> It's more directly for describing another object.
> 
> I'll take a look at `label` also. A literal for describing the content seems like a very generic term, so I think we should try to use an existing vocabulary before we define a new one.

Yeah, but from the [definition for `dc:subject`](https://www.dublincore.org/specifications/dublin-core/dcmi-terms/#http://purl.org/dc/terms/subject) it says:
> Recommended practice is to refer to the subject with a URI. If this is not possible or feasible, a literal value that identifies the subject may be provided. Both should preferably refer to a subject in a controlled vocabulary.

So to follow recommended practice you'd want a referenceable URI, that's why I'm not sure it makes much sense when we already have `as:summary` if we just want a string literal value.

On labels, they're an object type, much like a Hashtag but with more specific focus on moderation/content, we wouldn't be adding a `labels` property, just including items in `tags` that happen to be of `type` `Label` with their dereferenceable `id`.

We noted this decision around not introducing a explicit `contentWarning` property in the recent meeting notes, but I haven't had time to write the FEP that states our position.

The contents in a free text `contentWarning` attribute would be the same as that in a article's `summary` and best practice would be to include in the summary the content warnings, and to pair that up with labels via `tags`

@dariusk might be able to explain it better.

**Comment by evanp** [comment]
I think I described above how to use various properties for describing what about the content makes it potentially `sensitive`. I think `summary` is a good fallback if the other properties aren't available.

But `summary` is not only used for content warnings.

Especially if the content is traumatic, I think including showing the full `summary` to the user is poor first choice. A `summary` of an image of a spider would include a lot of description of the species, appearance, etc. of the spider, which might be upsetting for someone with arachnophobia. A plain text topic marker like "spider" would be a much better warning.

**Comment by dariusk** [comment]
I think I would rephrase what Emelia said here:

> the content warning text should actually be in the `summary`

In my mind, the `summary` is for a content _description_ and most automated filtering will ideally happen via a `label` system. If clients/servers decide to also regex `summary` for filtering, they can of course do that too.

And critically, to me a content _description_ is a form of content warning, because it is telling you what the content is but in less detail than in the content itself!

What we are actually trying to avoid here is free-text topic markers or the addition of a `contentWarning` text field. We would much rather have a `tag` with a `label` with `name: "spider"` (along with the other properties of a `label`) along with a `sensitive` flag. The `summary` could then include the species and other details because a person with arachnophobia would presumably have a rule to reliably filter anything associated with the spider label Object.

The desired behavior from clients, from our perspective, is to look at `sensitive` and/or the presence of `label` to figure out whether to collapse content. But `summary` remains a content _description_ field that some people will continue to put content _warnings_ into because it's free text and they can do what they want.

So I think this proposal actually is compatible with what you're saying, @evanp

**Comment by dariusk** [comment]
You can see more discussion in the May 13 2025 meeting starting here:

https://github.com/swicg/activitypub-trust-and-safety/blob/755498e3e04b5f970a84e1a56c157a622ec7c6ce/meetings/2025-05-13-taskforce-meeting.md?plain=1#L69

Specifically Jaz is saying

> I would love it if `summary` remained a summary and people can write whatever they want and that definitionally is not a content warning. Mastodon does the Hazard Tape and CW thing, but if we do this that could stop happen.

(that represents the general group consensus)

**Comment by evanp** [comment]
Another good property for describing text is [schema:keywords](https://schema.org/keywords), which is open to both URIs and plain text. That might be a good one to use.

**Comment by ThisIsMissEm** [comment]
Okay so final comment on this issue:

The taskforce won't be publishing a "Content Warnings" FEP (or if we did it would just be documenting `summary` used with `as:sensitive` which is the current status of content warnings in ActivityPub software).

Defining a new field for free text does not move us forwards, and has [too many migration problems](https://github.com/swicg/activitypub-trust-and-safety/issues/1#issuecomment-2769502144). This is the [group consensus](https://github.com/swicg/activitypub-trust-and-safety/issues/1#issuecomment-2959477995). A free text "content warning" does not empower users to filter content without relying on string matching (scunthorpe problem) and/or regular expressions (far too complex for the average person).

Instead, we see `summary` as being the summary of a object, which may include text that describes any content that may offend or harm. [Content Labels](https://github.com/swicg/activitypub-trust-and-safety/issues/84) are preferred instead, as this provides a structured way to have control over what content a user wants to see, removes the burden on the publisher to self-censor (via allowing third-party labellers), and does not rely on matching text in free text strings to function. We can also advertise vocabularies of labels, making it easier to integrate into user interfaces.

**Comment by trwnh** [comment]
> we see summary as being the summary of a object

this is fine

> [Content Labels](https://github.com/swicg/activitypub-trust-and-safety/issues/84) are preferred instead, as this provides a structured way to have control over what content a user wants to see, removes the burden on the publisher to self-censor (via allowing third-party labellers), and does not rely on matching text in free text strings to function.

this is also fine, but the publisher should be able to describe what is "safe" to show when as:sensitive is used. right now, the answer is "synthesize a safe preview from as:summary". this could be progressively enhanced by allowing publishers to explicitly denote "safe" information if as:summary is itself also "sensitive". it looks like evan proposes some kind of "sandbox" property in https://github.com/swicg/activitypub-trust-and-safety/issues/1#issuecomment-2968088006 which could work, although probably needs to be workshopped in a FEP or something first.

we could at the very least explore how to process as:sensitive in the initial workstream? like "here's what can be done with the existing as:sensitive and some potential future extensions" #32

**Comment by ThisIsMissEm** [comment]
I don't currently see how you could implement something to supersede `summary + as:sensitive`, and we'll create issues for the following explainer documents:
- Explainer on `as:sensitive` processing model - "Handling  sensitive content in ActivityPub"
- Explainer on Content Warnings as they currently exist, short and sweet, (`summary + as:sensitive = content warning`) and drawbacks identified — should link to the broader `as:sensitive` processing model above and probably to Content Labels FEP
